File: leetCode/string/Problem017.java
package leetCode.string;

import java.util.LinkedList;
import java.util.List;

/**
 * Given a string containing digits from 2-9 inclusive,
 * return all possible letter combinations that the number could represent.
 * A mapping of digit to letters (just like on the telephone buttons) is given below.
 * Note that 1 does not map to any letters.
 */
public class Problem017 {
    // iterative solution
    public List<String> letterCombinations(String digits) {
        LinkedList<String> result = new LinkedList<>();
        if (digits == null || digits.length() == 0) return result;
        result.add("");
        String[] letters = {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        for (int i = 0; i < digits.length(); ++i) {
            String letter = letters[digits.charAt(i) - '0'];
            while (result.peek().length() == i) {
                String removed = result.remove();
                for (char c : letter.toCharArray()) {
                    result.add(removed + c);
                }
            }
        }
        return result;
    }
}

File: leetCode/string/Problem017.cpp
#include <string>
#include <vector>

std::vector<std::string> mp = {"",    "",    "abc",  "def", "ghi",
                               "jkl", "mno", "pqrs", "tuv", "wxyz"};

// depth-first solution
void dfs(std::string&, int, std::string&, std::vector<std::string>&);
std::vector<std::string> letterCombinations1(std::string digits) {
  std::vector<std::string> res;
  std::string cur;
  dfs(digits, 0, cur, res);
  return res;
}

void dfs(std::string& digits, int idx, std::string& cur,
         std::vector<std::string>& res) {
  if (idx == digits.size()) {
    if (!cur.empty()) res.push_back(cur);
    return;
  }
  for (int i = 0; i < mp[digits[idx] - '0'].size(); ++i) {
    cur += mp[digits[idx] - '0'][i];
    dfs(digits, idx + 1, cur, res);
    cur.pop_back();
  }
}

// iterative solution
std::vector<std::string> letterCombinations2(std::string digits) {
  if (digits.empty()) return {};
  std::vector<std::string> result;
  result.push_back("");
  for (auto digit : digits) {
    std::vector<std::string> tmp;
    for (auto candidate : mp[digit - '0']) {
      for (auto s : result) {
        tmp.push_back(s + candidate);
      }
    }
    result.swap(tmp);
  }
  return result;
}